### PR TITLE
VP-5035: Improve notification migration v2->v3

### PR DESCRIPTION
**Pull request by akak1977**

### CODE REVIEW

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs [hunk 1]
```diff
@@ -224,15 +231,18 @@ FROM INFORMATION_SCHEMA.TABLES
 	                            null, null
                             FROM 
                             [PlatformNotificationTemplate] pnt
-                            INNER JOIN 
-                            (
-	                            SELECT [NotificationTypeId], MAX([Id]) [Id] FROM [PlatformNotificationTemplate]
-	                            GROUP BY [NotificationTypeId]
-                            ) pnt2
+                            INNER JOIN pnt2
```
Comment (kostyrin) [inline src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:234]: Why do you join pnt2 there ? I don't see link on pnt2 in the select

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs [hunk 2]
```diff
@@ -266,15 +276,33 @@ INNER JOIN
 			                            END
 	                            END
                             FROM [PlatformNotificationTemplate] pnt
-                            INNER JOIN 
+                            INNER JOIN pnt2
```
Comment (kostyrin) [inline src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:279]: We could lost data. e.g. Look at screen
image

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs [hunk 3]
```diff
@@ -322,16 +350,46 @@ INNER JOIN
     			                            THEN 'SmsNotificationMessageEntity'
     			                            ELSE 'EmailNotificationMessageEntity'
     		                            END
-	                            END
+	                            END [Discriminator]
 	                            FROM [PlatformNotification] pn
-	                            INNER JOIN 
+	                            INNER JOIN pnt2
```
Comment (kostyrin) [inline src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:355]: are there notification messages without NotificationId? Is it ok?
image

**Review by kostyrin — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs (hunk 3 above)
Comment (akak1977) [reply src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:355]: Yes, they are.

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs (hunk 2 above)
Comment (akak1977) [reply src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:279]: I'll look at this.

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs (hunk 1 above)
Comment (akak1977) [reply src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:234]: Just to define notification ID as MAX from template ids.

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs (hunk 3 above)
Comment (kostyrin) [reply src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:355]: did you check opening details in notification feeds?

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs (hunk 2 above)
Comment (akak1977) [reply src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:279]: I didn't touch this. It was made before. Looks it like made specially:
``` sql
    CASE WHEN pnt.[NotificationTypeId] LIKE '%EmailNotification%' THEN [Body] ELSE '' END,
	CASE WHEN pnt.[NotificationTypeId] LIKE '%SmsNotification%' THEN [Body] ELSE '' END,
```

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs (hunk 3 above)
Comment (akak1977) [reply src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:355]: Yes, I did.

File: src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs (hunk 2 above)
Comment (kostyrin) [reply src/VirtoCommerce.NotificationsModule.Data/Migrations/20190528144557_InitialNotifications.cs:279]: agree it's an old bug, maybe need to find solution

**Review by yecli — APPROVED** [review]

**Pull request merged**